### PR TITLE
Issue #12 Replace the MonadReader instance with a more general version

### PR DESCRIPTION
**Pull request by jfischoff**

### CODE REVIEW

File: src/Database/PostgreSQL/Transaction.hs [hunk 1]
```diff
@@ -60,9 +61,15 @@ newtype PGTransactionT m a =
              , Applicative
              , Monad
              , MonadTrans
-             , MonadReader Postgres.Connection
              , MonadIO
              )
+             
+instance MonadReader r m => MonadReader r (PGTransactionT m) where
+  ask = lift ask
+  local = undefined
```
Comment (patrickt) [inline src/Database/PostgreSQL/Transaction.hs]: What's going on here?

File: src/Database/PostgreSQL/Transaction.hs [hunk 2]
```diff
@@ -4,6 +4,7 @@
 {-# LANGUAGE GeneralizedNewtypeDeriving #-}
 {-# LANGUAGE ImplicitPrelude            #-}
 {-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE UndecidableInstances       #-}
```
Comment (patrickt) [inline src/Database/PostgreSQL/Transaction.hs]: I would prefer to leave UndecidableInstances out of this, if at all possible. Down there lies madness.

File: src/Database/PostgreSQL/Transaction.hs (hunk 1 above)
Comment (jfischoff) [inline src/Database/PostgreSQL/Transaction.hs]: Heh, sorry I left that in. I'll fix that when I get home tonight.

File: src/Database/PostgreSQL/Transaction.hs (hunk 2 above)
Comment (jfischoff) [inline src/Database/PostgreSQL/Transaction.hs]: UndecidableInstances are par for the course for the mtl instances. Look at the source of mtl for an explanation. 

https://hackage.haskell.org/package/mtl-2.2.1/docs/src/Control-Monad-Reader-Class.html

There isn't much madness to UndecideableInstances, perhaps you mean IncoherentInstances? 

UndecideableInstances don't lead to unpredictable instances, the compiler might hit a recursion limit, but I don't think that is known to happen with the way people use mtl.

File: src/Database/PostgreSQL/Transaction.hs [hunk 3]
```diff
@@ -60,9 +61,15 @@ newtype PGTransactionT m a =
...
-             , MonadReader Postgres.Connection
              , MonadIO
              )
+             
+instance MonadReader r m => MonadReader r (PGTransactionT m) where
+  ask = lift ask
+  local = undefined
+             
+getConnection :: Monad m => PGTransactionT m Postgres.Connection
+getConnection = PGTransactionT $ ReaderT $ return . id
```
Comment (lseppala) [inline src/Database/PostgreSQL/Transaction.hs]: Shorter implementation: `getConnection = PGTransactionT ask`

**Comment by reiddraper** [comment]
Thanks @jfischoff. Would you mind squashing this, and writing a clean commit message, once it's ready?

**Comment by jfischoff** [comment]
Sure

**Comment by jfischoff** [comment]
squashed

**Comment by reiddraper** [comment]
Thanks! Looks good to me. @lseppala, @patrickt, ready to merge?

**Pull request merged**